code/modules/xenoarchaeology/xenoartifact_console.dm

```diff
@@ -0,0 +1,256 @@
+				"name" = S.name,
+				"dialogue" = S.dialogue,
+				"price" = S.price,
+				"id" = S.unique_id,
+		))
+	data["buyer"] = list()
+	for(var/datum/xenoartifact_seller/buyer/B as() in buyers) //Buyer data
+		if(B)
+			data["buyer"] += list(list(
+				"name" = B.name,
+				"dialogue" = B.dialogue,
+				"price" = B.price,
+				"id" = B.unique_id,
+		))
+	data["tab_index"] = tab_index
+	data["current_tab"] = current_tab
+	data["tab_info"] = current_tab_info
+	data["linked_machines"] = linked_machines
+	data["sold_artifacts"] = sold_artifacts
+	return data
+
+/obj/machinery/computer/xenoartifact_console/ui_act(action, params) //I should probably use a switch statement for this but, the for statements look painful
+	. = TRUE
+	if(..())
+		return
+
+	if(action == "link_nearby")
+		sync_devices()
+		return
+	else if(action == "sell")
+		sell()
+		return
+
+	for(var/t in tab_index) //Handle swapping tabs.
+		if(action == "set_tab_[t]")
+			if(current_tab != t)
+				current_tab = t
+				switch(t)
+					if("Listings")//Not the best way of doing this but I can't be fucked otherwise.
+						current_tab_info = "Here you can find listings for various research samples, usually fresh from the field. These samples aren't distrubuted by the Nanotrasen affiliated cargo system, so instead listing data is sourced from stray bluespace-threads."
+					if("Export")
+						current_tab_info = "Sell any export your department produces through open bluespace strings. Anonymously trade and sell ancient alien bombs, explosive slime cores, or just regular bombs."
+					if("Linking")
+						current_tab_info = "Link machines to the Listing Console."
+			else if(current_tab == t)
+				current_tab = ""
+				current_tab_info = ""
+			return
+
+	for(var/datum/xenoartifact_seller/S as() in sellers) //Handle buying artifacts.
+		if(action == "purchase_[S.unique_id]")
+			if(!linked_inbox)
+				say("Error. No linked hardware.")
+			else if(budget.account_balance-S.price < 0)
+				say("Error. Insufficient funds.")
+			else if(linked_inbox && budget.account_balance-S.price >= 0)
+				var/obj/item/xenoartifact/A = new (get_turf(linked_inbox.loc), S.difficulty)
+				var/datum/component/xenoartifact_pricing/X = A.GetComponent(/datum/component/xenoartifact_pricing)
+				if(X)
+					X.price = S.price
+					sellers -= S
+					budget.adjust_money(-1*S.price)
+					say("Purchase complete. [budget.account_balance] credits remaining in Research Budget")
+					addtimer(CALLBACK(src, .proc/generate_new_seller), (rand(1,5)*60) SECONDS)
+					A = null
+
+	update_icon()
+
+/obj/machinery/computer/xenoartifact_console/proc/sell()
+	if(!linked_inbox)
+		say("Error. No linked hardware.")
+	else
+		var/obj/selling_item
+		for(var/obj/I in oview(1, linked_inbox))
+			for(var/datum/xenoartifact_seller/buyer/B in buyers)
+				if(istype(I, B.buying))
+					buyers -= B
+					addtimer(CALLBACK(src, .proc/generate_new_buyer), (rand(1,5)*60) SECONDS)
+					selling_item = I
+					break
```

Comment:
this is very weird way of selling something.
Why are you looping the buyers? Can't you send a ref to a buyer of choice?

Reply:
This makes selling things a bit more leisurely. You can just choose to export something and have the console handle finding a buyer for the item.